use "custom" folder instead of "download"

eisbaer66 opened the pull request:
seperates the assetpack cleanly from other downloads
prevents deleting otherwise downloaded and maybe still needed files

I feel like this should be the default. 
If installing to the "download" folder needs to be supported, I'd suggest to at least giving the user an option to use the "custom" folder instead.

rob5300 commented:
This may be simple to implement and would be beneficial long term. Will keep note to do this in future.

rob5300 commented:
Just checking over this, changes look good.
However you shouldn't overstep the purpose of this PR and also remove usage of promises and change to async await. This is a good change but not relevant to this PR. I think its acceptable for new functions but not those that do not need change for this change to work.
Also the addition of some ts type defs is nice but again is not relevant seeing as its not used in the whole project.

Do tell me what you think of this. For me to accept this I would wish it to fulfill the purpose it originally specified.

eisbaer66 commented:
### async/await
I rewrote the file_manger.SaveFileList-method because 1. it wasn't working properly (wasn't stringifing the filelist) and 2. it wasn't used. At that point I thought I treat it like a new function and redo it the nice way ;)
But I can reset the mod_manager.Setup function and "upward" to use promises again (to match the previous code). That part is truly out of scope.
I'd want to keep my handleMajorError function in the main.js, though, since it logs the error on top of keeping you from repeating yourself or do you want a separate PR for that?
### typedefs
I added JSDoc-comments to the new methods. It helps giving meaningful tooltips in your editor . If you wanna reuse a typedef in multiple files you have to put it somewhere where both files can get the typedef. That's what the typedefs.js is for. Others don't need to use it to work, but are free to of cause. So i thought it's ok to add, since keeps us from having to keep multiple typedefs up-to-date if they change.
If you don't want to make them mandatory, that's ok with me, but for now having them is better the not having them ;)
I feel like they should be allowed for new methods, at least. But adding comments and typedefs for existing stuff could be it's own issue again.
Is this ok to keep? or do you want it changed?

rob5300 commented on modules/config.js:
```diff
@@ -125,6 +126,25 @@ module.exports = {
+     * @param {string} p path tp expand
+     * @returns {string} replaced and normalized path
+     */
+    ExpandVariables(p){
```
Could maybe have a name that better describes whats happening as "expanding" is often used for macros.
Try ResolveVariables or FormatVariables.

eisbaer66 replied:
mostly inspired by the Environment.ExpandEnvironmentVariables in .net ^^
will change to ResolveVariables

rob5300 replied:
That's true, but when making this i was think more of string.format from .NET. Shame none keep it standard

eisbaer66 commented on modules/file_manager.js:
```diff
@@ -64,6 +64,66 @@ GetPath(){
\ No newline at end of file
+function IsNotFoundError(e) {
+    return e.code === "ENOENT";
+}
```
Oh file_manager isn't meant to be for files in general but for the files of mods? 
Maybe mod_file_manager would be a better name? or just mod_files? Since I like to avoid words like "manager" (they dont really add information to the class; you can add "Manager" to most class/module-name without changing it's meaning, they all are managing something ^^)
But I'm getting side tracked ^^ you have a good idea for the new modulename? 
Maybe these new functions shouldn't  even be in a new module, but extend the build in fs module? so you can do the build in global.fsPromises.mkdir(...), global.fsPromises.rename(...), etc. as well as the new global.fsPromises.move(...), global.fsPromises.fileExists(...), etc.?
Then "fs_extensions" could be a good name?

rob5300 replied:
Yea i think an extension to fs would be good. Do also try to share requires when you can like i am via global to prevent wasteful loading.

eisbaer66 commented on modules/file_manager.js:
```diff
@@ -39,12 +42,9 @@ GetFileListSync(modName){
-        fs.writeFileSync(path, filelist, "utf8");
-        resolve();
-    });
+async SaveFileList(filelist, modName){
```
will still have to add the JSON.stringify(filelist)
but i will undo the async and use of fsPromise

rob5300 commented on main.js:
```diff
@@ -59,41 +59,24 @@ function createWindow() {
-                    app.quit();
-                });
-            }
+            return mod_manager.Setup().then(() => {
```
Because migration happens on mod manager setup, this holds the window from loading its page. Please move the calling of this to after the window page is loaded.

eisbaer66 commented on modules/mod_manager.js:
```diff
@@ -48,6 +50,7 @@ module.exports =
 
 //Export variables.
 all_mods_data: null,
+modDefinitions: {"_blank": {"name":"_blank","version":0.1}},
```
Gonna remove GetModVersionInfos() in config.js, too.
It's just there to filter out the _blank mod

eisbaer66 replied:
just realized removing the function is probably a bad idea since it could still be in user config
so I'm keeping the function it for now

rob5300 commented:
New issue I found which may not allow the use of custom. The client will try to redownload all the game data from fast dl if their download filter is all, making this not ideal. If we can have a sub folder in download that would be better.

eisbaer66 commented:
> New issue I found which may not allow the use of custom. The client will try to redownload all the game data from fast dl if their download filter is all, making this not ideal. If we can have a sub folder in download that would be better.

I don't think a subfolder is possible in the download-folder. It has a fixed structure. I'm pretty sure that's what the custom folder is for. You can make a subfolder in custom and then put anything you can put in download in this subfolder. To the game it should make no difference if its under "download" or "custom/AwesomeModName".

```
\materials\models\weapons\c_items\c_kanopykiller.vmt
\materials\models\weapons\c_items\c_kanopykiller_normal.vtf
\materials\models\weapons\c_items\c_lead_blaster.vmt
\materials\models\weapons\c_items\c_lead_blaster_blue.vmt
```
These files get downloaded for me. But they are not part of the assetpack-zip as far as I can tell.
Do you see any other files getting downloaded?
I don't think this is a general problem with the custom folder ... for now ^^

rob5300 commented:
Going to update here as this may still be a good addition.
Once the new depot system is merged in it would be best to merge to update and re check if these changes are still feasible.
Due to lack of time this may take a while but is something I am slowly doing.

rob5300 commented:
I have an update.
I recently added the Events content and that is using custom.
This however causes issues where content can easily get duplicated if it was already in /download/
I am reluctant now to have our main content also go there, and i plan to move the events content back to /download/ due to this problem.

This would make this not needed anymore, but well I wouldnt want code to get wasted. Hopefully some parts here can be kept and merged.
Feel free to DM me as I have had little time recently to work on things here so I certainly welcome help in other places if you wanted to. I have started to use TS on new features so that is an avenue to explore more.
